ow_lander/scripts/arm_action_servers.py

```diff
@@ -0,0 +1,666 @@
+    def on_DigCircular_action(self,goal):
+        plan = self._update_motion(goal)
+        if self.current_traj == False: 
+            self._server.set_aborted(self._result)
+            return 
+        success = False
+
+        self.trajectory_async_executer.execute(self.current_traj.joint_trajectory,
+                                           done_cb=None,
+                                           active_cb=None,
+                                           feedback_cb=self.trajectory_async_executer.stop_arm_if_fault)
+
+        # Record start time
+        start_time = rospy.get_time()
+
+        def now_from_start(start):
+            return rospy.Duration(secs=rospy.get_time() - start)
+
+        while ((now_from_start(start_time) < self._timeout)):
+
+           self._update_feedback()
+           
+        success = self.trajectory_async_executer.success() and self.trajectory_async_executer.wait()        
+        
+            
+        if success:
+            self._result.final.x = self._fdbk.current.x
+            self._result.final.y = self._fdbk.current.y 
+            self._result.final.z = self._fdbk.current.z 
+            rospy.loginfo('%s: Succeeded' % self._action_name)
+            self._server.set_succeeded(self._result)
+        else:
+            rospy.loginfo('%s: Failed' % self._action_name)
+            self._server.set_aborted(self._result)         
+
+
+class DigLinearActionServer(object):
+    
+    def __init__(self,name):
+        self._action_name = name
+        self._server = actionlib.SimpleActionServer(self._action_name, 
+                                                    ow_lander.msg.DigLinearAction, 
+                                                    execute_cb=self.on_DigLinear_action, 
+                                                    auto_start = False)
+        self._server.start()
+        # Action Feedback/Result
+        self._fdbk = ow_lander.msg.DigLinearFeedback()
+        self._result = ow_lander.msg.DigLinearResult()
+        self._current_link_state = LinkStateSubscriber()
+        self._interface = MoveItInterface()
+        self._timeout = 0.0
+        self.trajectory_async_executer = TrajectoryAsyncExecuter()
+        self.trajectory_async_executer.connect("arm_controller")
+        self.current_traj = RobotTrajectory()
+        
+    
+    def _update_feedback(self):
+ 
+        self._ls =  self._current_link_state._link_value
+        self._fdbk.current.x = self._ls.x
+        self._fdbk.current.y = self._ls.y
+        self._fdbk.current.z = self._ls.z
+        self._server.publish_feedback(self._fdbk)
+
+        
+        
+    def _update_motion(self, goal):
+        
+        print("DigLinear activity started")
+        self.current_traj  = all_action_trajectories.dig_linear(self._interface.move_arm,
+                                                          self._interface.robot,  
+                                                          self._interface.moveit_fk, goal)
```

Comment:
Indent.

Reply:
downloaded python suggested app to auto -intend all in vs code... should fix all